### PR TITLE
Deploy MetaSwap contract with a new library and 3 new meta pools

### PR DESCRIPTION
**Pull request by penandlim**
# Deploy MetaSwap contract with a new library and 3 new meta pools

## Description

Deploys 3 new meta pools with a new MetaSwapUtils library.

Fixes https://github.com/saddle-finance/tickets/issues/307

## Deployment verfication checklist

List any new deployments that is expected to be saved after running the deploy task. After running the scripts, check off each item if they are successfully deployed.

Please provide explorer links to the deployed contracts and verify them using either standard solcInputs json or manually flattened contracts using `npx hardhat flatten`. After verifying them, check off each item.

Manually created
- [x] MetaSwapDeposit (same deployment as existing `SaddleSUSDMetaPoolDeposit.json`)

Expected to be created by the deploy scripts
- [x] MetaSwapUtils https://etherscan.io/address/0x505736d427C313CdeCdbf72049Eb944e53B3065b#code
- [x] MetaSwapV3 https://etherscan.io/address/0x4568727f50c7246ded8C39214Ed6FF3c157f080D#code
- [x] SaddleSUSDMetaPoolV3 https://etherscan.io/address/0x4568727f50c7246ded8C39214Ed6FF3c157f080D#code
- [x] SaddleSUSDMetaPoolV3LPToken https://etherscan.io/address/0x444F94460a641429CDa4e38E02E51642Cc38276A#code
- [x] SaddleSUSDMetaPoolV3Deposit  https://etherscan.io/address/0xB98fd1f66884cD5786b37cDE040B9f0cf763866f#code
- [x] SaddleTBTCMetaPoolV3 https://etherscan.io/address/0xfa9ED0309Bf79Eb84C847819F0B3CB84F6d351Af#code
- [x] SaddleTBTCMetaPoolV3LPToken  https://etherscan.io/address/0xA2E81Eb93F0F9814ae9A3bea2D2A63408f2709C1#code
- [x] SaddleTBTCMetaPoolV3Deposit https://etherscan.io/address/0x4946DE721ce70D4B7aa226aA0Fe869C935769388#code
- [x] SaddleWCUSDMetaPoolV3 https://etherscan.io/address/0xB62222B941e9B652BE3632EEa062cb0ff66b1d1c#code
- [x] SaddleWCUSDMetaPoolV3LPToken https://etherscan.io/address/0x0dB8b09c13FE21913faF463274cE8e0a51719f16#code
- [x] SaddleWCUSDMetaPoolV3Deposit https://etherscan.io/address/0x671D5942F901F5C60e4EbaD1c3bF284A4d28c675#code

```
┌─────────┬────────────────────────────────┬──────────────────────────────────────────────┐
│ (index) │               0                │                      1                       │
├─────────┼────────────────────────────────┼──────────────────────────────────────────────┤
│    0    │     'SaddleSUSDMetaPoolV3'     │ '0x4568727f50c7246ded8C39214Ed6FF3c157f080D' │
│    1    │          'MetaSwapV3'          │ '0x4568727f50c7246ded8C39214Ed6FF3c157f080D' │
│    2    │ 'SaddleSUSDMetaPoolV3LPToken'  │ '0x444F94460a641429CDa4e38E02E51642Cc38276A' │
│    3    │ 'SaddleSUSDMetaPoolV3Deposit'  │ '0xB98fd1f66884cD5786b37cDE040B9f0cf763866f' │
│    4    │     'SaddleTBTCMetaPoolV3'     │ '0xfa9ED0309Bf79Eb84C847819F0B3CB84F6d351Af' │
│    5    │ 'SaddleTBTCMetaPoolV3LPToken'  │ '0xA2E81Eb93F0F9814ae9A3bea2D2A63408f2709C1' │
│    6    │ 'SaddleTBTCMetaPoolV3Deposit'  │ '0x4946DE721ce70D4B7aa226aA0Fe869C935769388' │
│    7    │    'SaddleWCUSDMetaPoolV3'     │ '0xB62222B941e9B652BE3632EEa062cb0ff66b1d1c' │
│    8    │ 'SaddleWCUSDMetaPoolV3LPToken' │ '0x0dB8b09c13FE21913faF463274cE8e0a51719f16' │
│    9    │ 'SaddleWCUSDMetaPoolV3Deposit' │ '0x671D5942F901F5C60e4EbaD1c3bF284A4d28c675' │
└─────────┴────────────────────────────────┴──────────────────────────────────────────────┘
```

## Checklist:

- [x] Any imported tokens's addresses are verified against official docs
- [x] If any existing contracts are changed, ensure the saved deployments are removed before running the deploy scripts
- [x] If deploying a new pool on a testnet, ensure there is liquidity by providing the test tokens to a DEX.
- [x] If possible, the deploy scripts are tested against forked mainnet or local hardhat node.
- [x] Newly deployed contracts match the expected list of new deployments
- [x] All newly deployed contracts are verified on explorers

### CODE REVIEW

File: deploy/mainnet/460_deploy_SUSDMetaPoolV3.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+const BASE_POOL_NAME = "SaddleUSDPoolV2"
+
+// Constructor arguments
+const TOKEN_NAMES = ["SUSD", `${BASE_POOL_NAME}LPToken`]
+const TOKEN_DECIMALS = [18, 18]
+const LP_TOKEN_NAME = "Saddle sUSD/saddleUSD-V2"
+const LP_TOKEN_SYMBOL = "saddleSUSD"
+const INITIAL_A = 100
+const SWAP_FEE = 4e6 // 4bps
+const ADMIN_FEE = 0
```
Comment (mjdatadog) [inline deploy/mainnet/460_deploy_SUSDMetaPoolV3.ts:17]: Do we want to initialize with an admin fee?

File: deploy/mainnet/460_deploy_SUSDMetaPoolV3.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+// Deployment names
+const META_POOL_NAME = "SaddleSUSDMetaPoolV3"
+const META_POOL_LP_TOKEN_NAME = `${META_POOL_NAME}LPToken`
+const BASE_POOL_NAME = "SaddleUSDPoolV2"
+
+// Constructor arguments
+const TOKEN_NAMES = ["SUSD", `${BASE_POOL_NAME}LPToken`]
+const TOKEN_DECIMALS = [18, 18]
+const LP_TOKEN_NAME = "Saddle sUSD/saddleUSD-V2"
```
Comment (mjdatadog) [inline deploy/mainnet/460_deploy_SUSDMetaPoolV3.ts:13]: I believe this should be V3 correct? this same LP token name was deploy with the SUSD 'Updated' Metapool

File: deploy/mainnet/462_deploy_TBTCMetaPoolV3.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+// Deployment names
+const META_POOL_NAME = "SaddleTBTCMetaPoolV3"
+const META_POOL_LP_TOKEN_NAME = `${META_POOL_NAME}LPToken`
+const BASE_POOL_NAME = "SaddleBTCPoolV2"
+
+// Constructor arguments
+const TOKEN_NAMES = ["TBTCv2", `${BASE_POOL_NAME}LPToken`]
+const TOKEN_DECIMALS = [18, 18]
+const LP_TOKEN_NAME = "Saddle tBTCv2/saddleWRenSBTC"
```
Comment (mjdatadog) [inline deploy/mainnet/462_deploy_TBTCMetaPoolV3.ts:12]: Ditto on comment about V3

File: deploy/mainnet/465_deploy_CUSDMetaPoolV3Deposit.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      ).address,
+      (
+        await get(META_POOL_LP_TOKEN_NAME)
+      ).address,
+    )
+  }
+}
+export default func
+func.tags = [META_POOL_DEPOSIT_NAME]
+func.dependencies = ["SUSDMetaPoolTokens", META_POOL_NAME]
```
Comment (mjdatadog) [inline deploy/mainnet/465_deploy_CUSDMetaPoolV3Deposit.ts:63]: Another dependency change

File: deploy/mainnet/464_deploy_CUSDMetaPoolV3.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+// Deployment names
+const META_POOL_NAME = "SaddleWCUSDMetaPoolV3"
+const META_POOL_LP_TOKEN_NAME = `${META_POOL_NAME}LPToken`
+const BASE_POOL_NAME = "SaddleUSDPoolV2"
+
+// Constructor arguments
+const TOKEN_NAMES = ["WCUSD", `${BASE_POOL_NAME}LPToken`]
+const TOKEN_DECIMALS = [18, 18]
+const LP_TOKEN_NAME = "Saddle wCUSD/saddleUSD-V2"
```
Comment (mjdatadog) [inline deploy/mainnet/464_deploy_CUSDMetaPoolV3.ts:12]: V3 as well

**Comment by penandlim** [comment]
@nguyenaiden 
1. After MetaSwapUtils is deployed, we can verify it on etherscan with our soure code to ensure the correct version was deployed.
2. After MetaSwap is deployed, verify on etherscan. Then use etherscan script to check correct library was used.
3. After each meta pools are initialized, we can use script from #577 on a forked network and check if the deployed instance will suffer any virtual price change.

File: deploy/mainnet/465_deploy_CUSDMetaPoolV3Deposit.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        log: true,
+      },
+      "clone",
+      (
+        await get(TARGET_META_SWAP_DEPOSIT_NAME)
+      ).address,
+    )
+
+    const newPoolEvent = receipt?.events?.find(
+      (e: any) => e["event"] == "NewClone",
```
Comment (ug02fast) [inline deploy/mainnet/465_deploy_CUSDMetaPoolV3Deposit.ts:34]: +1 on Duy's comments. Should never use loose equality in JS. https://2ality.com/2011/12/strict-equality-exemptions.html

File: deploy/mainnet/182_deploy_SUSDMetaPoolDeposit.ts [hunk 7]
```diff
@@ -32,6 +32,10 @@ const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
         await get("SaddleSUSDMetaPoolLPToken")
       ).address,
     )
+
+    if ((await getOrNull("MetaSwapDeposit")) == null) {
```
Comment (ug02fast) [inline deploy/mainnet/182_deploy_SUSDMetaPoolDeposit.ts:36]: +1 on Duy's comments. Should never use loose equality in JS. https://2ality.com/2011/12/strict-equality-exemptions.html

File: deploy/mainnet/463_deploy_TBTCMetaPoolV3Deposit.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        log: true,
+      },
+      "clone",
+      (
+        await get(TARGET_META_SWAP_DEPOSIT_NAME)
+      ).address,
+    )
+
+    const newPoolEvent = receipt?.events?.find(
+      (e: any) => e["event"] == "NewClone",
```
Comment (ug02fast) [inline deploy/mainnet/463_deploy_TBTCMetaPoolV3Deposit.ts:34]: +1 on Duy's comments. Should never use loose equality in JS. https://2ality.com/2011/12/strict-equality-exemptions.html

File: deploy/mainnet/464_deploy_CUSDMetaPoolV3.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+      (
+        await get("LPToken")
+      ).address,
+      (
+        await get(BASE_POOL_NAME)
+      ).address,
+    )
+
+    const newPoolEvent = receipt?.events?.find(
+      (e: any) => e["event"] == "NewSwapPool",
```
Comment (ug02fast) [inline deploy/mainnet/464_deploy_CUSDMetaPoolV3.ts:58]: +1 on Duy's comments. Should never use loose equality in JS. https://2ality.com/2011/12/strict-equality-exemptions.html

File: deploy/mainnet/467_add_to_GeneralizedSwapMigrator.ts [hunk 10]
```diff
@@ -0,0 +1,95 @@
+import { DeployFunction } from "hardhat-deploy/types"
+import { HardhatRuntimeEnvironment } from "hardhat/types"
+import { CHAIN_ID } from "../../utils/network"
+
+const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
+  const { deployments, getNamedAccounts, getChainId, ethers } = hre
+  const { execute, deploy, get, getOrNull, log, save } = deployments
+  const { deployer } = await getNamedAccounts()
+
+  if ((await getChainId()) == CHAIN_ID.MAINNET) {
```
Comment (ug02fast) [inline deploy/mainnet/467_add_to_GeneralizedSwapMigrator.ts:10]: +1 on Duy's comments. Should never use loose equality in JS. https://2ality.com/2011/12/strict-equality-exemptions.html

**Review by ug02fast — APPROVED** [review]

**Comment by penandlim** [comment]
```
New deployments:
┌─────────┬────────────────────────────────┬──────────────────────────────────────────────┐
│ (index) │               0                │                      1                       │
├─────────┼────────────────────────────────┼──────────────────────────────────────────────┤
│    0    │     'SaddleSUSDMetaPoolV3'     │ '0x4568727f50c7246ded8C39214Ed6FF3c157f080D' │
│    1    │          'MetaSwapV3'          │ '0x4568727f50c7246ded8C39214Ed6FF3c157f080D' │
│    2    │ 'SaddleSUSDMetaPoolV3LPToken'  │ '0x444F94460a641429CDa4e38E02E51642Cc38276A' │
│    3    │ 'SaddleSUSDMetaPoolV3Deposit'  │ '0xB98fd1f66884cD5786b37cDE040B9f0cf763866f' │
│    4    │     'SaddleTBTCMetaPoolV3'     │ '0xfa9ED0309Bf79Eb84C847819F0B3CB84F6d351Af' │
│    5    │ 'SaddleTBTCMetaPoolV3LPToken'  │ '0xA2E81Eb93F0F9814ae9A3bea2D2A63408f2709C1' │
│    6    │ 'SaddleTBTCMetaPoolV3Deposit'  │ '0x4946DE721ce70D4B7aa226aA0Fe869C935769388' │
│    7    │    'SaddleWCUSDMetaPoolV3'     │ '0xB62222B941e9B652BE3632EEa062cb0ff66b1d1c' │
│    8    │ 'SaddleWCUSDMetaPoolV3LPToken' │ '0x0dB8b09c13FE21913faF463274cE8e0a51719f16' │
│    9    │ 'SaddleWCUSDMetaPoolV3Deposit' │ '0x671D5942F901F5C60e4EbaD1c3bF284A4d28c675' │
└─────────┴────────────────────────────────┴──────────────────────────────────────────────┘

MetaSwapUtils : 0x505736d427C313CdeCdbf72049Eb944e53B3065b (deployed before deployer was fully replenished)
```

File: deploy/mainnet/466_register_new_pools.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    {
+      // tBTCv2 meta pool v3
+      poolAddress: (await get("SaddleTBTCMetaPoolV3")).address,
+      typeOfAsset: PoolType.BTC,
+      poolName: ethers.utils.formatBytes32String("tBTCv2-BTCv2_v3"),
+      targetAddress: (await get("MetaSwapV3")).address,
+      metaSwapDepositAddress: (await get("SaddleTBTCMetaPoolV3Deposit"))
+        .address,
+      isSaddleApproved: true,
+      isRemoved: true,
```
Comment (hammeiam) [inline deploy/mainnet/466_register_new_pools.ts:36]: why true?

File: deploy/mainnet/466_register_new_pools.ts [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+      log("Adding pools to registry")
+
+      const batchCall = await Promise.all(
+        pools.map(
+          async (pool) => await poolRegistry.populateTransaction.addPool(pool),
+        ),
+      )
+      const batchCallData = batchCall
+        .map((x) => x.data)
+        .filter((x): x is string => !!x)
```
Comment (hammeiam) [inline deploy/mainnet/466_register_new_pools.ts:68]: also `filter(Boolean)` is a cool trick that does the same thing

File: deployments/mainnet/SaddleALETHPoolLPToken.json [hunk 13]
```diff
@@ -222,5 +222,5 @@
       "type": "event"
     }
   ],
-  "numDeployments": 3
+  "numDeployments": 5
```
Comment (hammeiam) [inline deployments/mainnet/SaddleALETHPoolLPToken.json:225]: is this okay?

**Review by hammeiam — DISMISSED** [review]

File: deploy/mainnet/460_deploy_SUSDMetaPoolV3.ts [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    )
+
+    await deploy(META_POOL_NAME, {
+      from: deployer,
+      log: true,
+      contract: "MetaSwap",
+      skipIfAlreadyDeployed: true,
+      libraries: {
+        SwapUtils: (await get("SwapUtils")).address,
+        MetaSwapUtils: (await get("MetaSwapUtils")).address,
```
[the remark this answers is not in the document]
Comment (penandlim) [reply deploy/mainnet/460_deploy_SUSDMetaPoolV3.ts:40]: The only bad MetaSwapUtils was `deployments/mainnet/MetaSwapUtils.json`. 

And now that the old deployment json is removed, there isn't a way to get bad meta swap utils on any chain (unless manually getting the library from bad pools) so `MetaSwapUtils` will always refer to the correct deployments.

Going forward when making any changes to any libraries, we can append versions to the contract names for easier differentiation.

File: deploy/mainnet/466_register_new_pools.ts (hunk 11 above)
Comment (penandlim) [reply deploy/mainnet/466_register_new_pools.ts:36]: Good catch. That should be false.

File: deployments/mainnet/SaddleALETHPoolLPToken.json (hunk 13 above)
Comment (penandlim) [reply deployments/mainnet/SaddleALETHPoolLPToken.json:225]: Yeah its known behavior. We can fix this in another PR like this

https://github.com/saddle-finance/saddle-contract/pull/563/files

**Review by hammeiam — APPROVED** [review]

**Pull request merged**